File: integration-examples/axe-hmac/send_signed_alert.py
"""Envio de um alerta ALRT→AXE com assinatura HMAC-SHA256 do corpo bruto."""

import hashlib
import hmac
import json
import os
import sys
import urllib.error
import urllib.request
import uuid
from datetime import datetime, timezone

endpoint = os.getenv("AXE_URL", "https://dispatchapp-dmbshjft.manus.space/api/integrations/alrt/events")
api_key = os.getenv("AXE_API_KEY")
hmac_secret = os.getenv("AXE_HMAC_SECRET")

if not api_key or not hmac_secret:
    raise RuntimeError("Defina AXE_API_KEY e AXE_HMAC_SECRET no ambiente antes de executar.")

timestamp = datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")
external_id = os.getenv("ALERT_EXTERNAL_ID", f"demo-{uuid.uuid4()}")
correlation_id = str(uuid.uuid4())
payload = {
    "schemaVersion": "1.0",
    "eventId": f"evt_alrt_{uuid.uuid4()}",
    "eventType": "alert.received",
    "occurredAt": timestamp,
    "source": {"system": "despacho-alrt", "environment": "homologacao"},
    "correlationId": correlation_id,
    "idempotencyKey": f"alrt:alert:{external_id}:v1",
    "data": {
        "alert": {
            "externalId": external_id,
            "category": "Alerta urbano",
            "priority": "alta",
            "description": "Alerta de homologação assinado pelo ALRT.",
            "address": "Rua de Homologação, nº 100",
            "latitude": -27.0976,
            "longitude": -48.9104,
            "reportedAt": timestamp,
            "sourceStatus": "novo",
        }
    },
}

# serialização compacta e única: o mesmo conteúdo precisa ser assinado e enviado.
raw_body = json.dumps(payload, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
signed_value = timestamp.encode("utf-8") + b"." + raw_body
signature = "sha256=" + hmac.new(hmac_secret.encode("utf-8"), signed_value, hashlib.sha256).hexdigest()

request = urllib.request.Request(
    endpoint,
    data=raw_body,
    method="POST",
    headers={
        "Content-Type": "application/json",
        "X-ALRT-API-Key": api_key,
        "X-Timestamp": timestamp,
        "X-Request-Timestamp": timestamp,
        "X-Correlation-Id": correlation_id,
        "X-Signature": signature,
    },
)

try:
    with urllib.request.urlopen(request, timeout=15) as response:
        print(json.dumps({"status": response.status, "correlationId": correlation_id, "response": response.read().decode("utf-8")}, indent=2))
except urllib.error.HTTPError as error:
    print(json.dumps({"status": error.code, "correlationId": correlation_id, "response": error.read().decode("utf-8")}, indent=2), file=sys.stderr)
    raise SystemExit(1) from error

File: integration-examples/axe-hmac/send-signed-alert.mjs
import { createHmac, randomUUID } from "node:crypto";

const endpoint = process.env.AXE_URL ?? "https://dispatchapp-dmbshjft.manus.space/api/integrations/alrt/events";
const apiKey = process.env.AXE_API_KEY;
const hmacSecret = process.env.AXE_HMAC_SECRET;

if (!apiKey || !hmacSecret) {
  throw new Error("Defina AXE_API_KEY e AXE_HMAC_SECRET no ambiente antes de executar.");
}

const timestamp = new Date().toISOString();
const externalId = process.env.ALERT_EXTERNAL_ID ?? `demo-${Date.now()}`;
const correlationId = randomUUID();
const payload = {
  schemaVersion: "1.0",
  eventId: `evt_alrt_${randomUUID()}`,
  eventType: "alert.received",
  occurredAt: timestamp,
  source: { system: "despacho-alrt", environment: "homologacao" },
  correlationId,
  idempotencyKey: `alrt:alert:${externalId}:v1`,
  data: {
    alert: {
      externalId,
      category: "Alerta urbano",
      priority: "alta",
      description: "Alerta de homologação assinado pelo ALRT.",
      address: "Rua de Homologação, nº 100",
      latitude: -27.0976,
      longitude: -48.9104,
      reportedAt: timestamp,
      sourceStatus: "novo",
    },
  },
};

// Não reformate nem gere novamente o JSON depois desta linha: estes bytes são assinados e enviados.
const rawBody = JSON.stringify(payload);
const signature = `sha256=${createHmac("sha256", hmacSecret).update(`${timestamp}.${rawBody}`, "utf8").digest("hex")}`;
const controller = new AbortController();
const timeout = setTimeout(() => controller.abort(), 15_000);

try {
  const response = await fetch(endpoint, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-alrt-api-key": apiKey,
      "x-timestamp": timestamp,
      "x-request-timestamp": timestamp,
      "x-correlation-id": correlationId,
      "x-signature": signature,
    },
    body: rawBody,
    signal: controller.signal,
  });
  const responseBody = await response.text();
  console.log(JSON.stringify({ status: response.status, correlationId, response: responseBody }, null, 2));
  process.exitCode = response.ok ? 0 : 1;
} finally {
  clearTimeout(timeout);
}
